File: dotnet/src/SemanticKernel/Planning/PlanRunner.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Text;
using System.Threading.Tasks;
using System.Xml;
using Microsoft.Extensions.Logging;
using Microsoft.SemanticKernel.Orchestration;

namespace Microsoft.SemanticKernel.Planning;

/// <summary>
/// Executes plans created by the PlannerSkill/ProblemSolver semantic function.
/// </summary>
internal class PlanRunner
{
    /// <summary>
    /// The tag name used in the plan xml for the user's goal/ask.
    /// </summary>
    private const string GoalTag = "goal";

    /// <summary>
    /// The tag name used in the plan xml for the solution.
    /// </summary>
    private const string SolutionTag = "solution";

    /// <summary>
    /// The tag name used in the plan xml for a generic step.
    /// </summary>
    private const string StepTag = "step";

    /// <summary>
    /// The tag name used in the plan xml for the context variables.
    /// </summary>
    private const string VariablesTag = "variables";

    private readonly IKernel _kernel;

    public PlanRunner(IKernel kernel)
    {
        this._kernel = kernel;
    }

    /// <summary>
    /// Executes the next step of a plan xml.
    /// </summary>
    /// <param name="context">The context to execute the plan in.</param>
    /// <param name="planPayload">The plan xml.</param>
    /// <param name="defaultStepExecutor">The default step executor to use if a step does not have a skill function.</param>
    /// <returns>The resulting plan xml after executing a step in the plan.</returns>
    /// <context>
    /// Brief overview of how it works:
    /// 1. The plan xml is parsed into a list of goals, variables, arguments, and steps.
    /// 2. The first step is executed.
    /// 3. The resulting context is converted into a new plan xml.
    /// 4. The new plan xml is returned.
    /// </context>
    /// <exception cref="PlanningException">Thrown when the plan xml is invalid.</exception>
    public async Task<SKContext> ExecuteXmlPlanAsync(SKContext context, string planPayload, ISKFunction defaultStepExecutor)
    {
        try
        {
            XmlDocument xmlDoc = new();
            try
            {
                xmlDoc.LoadXml("<xml>" + planPayload + "</xml>");
            }
            catch (XmlException e)
            {
                throw new PlanningException(PlanningException.ErrorCodes.InvalidPlan, "Failed to parse plan xml.", e);
            }

            // Get the Goal
            var (_, goalXmlString) = GatherGoal(xmlDoc);

            // Get the <see cref="ContextVariables"/> and Solution
            XmlNodeList variables = xmlDoc.GetElementsByTagName(VariablesTag);
            XmlNodeList solution = xmlDoc.GetElementsByTagName(SolutionTag);

            // Prepare content for the new plan xml
            var variablesContent = new StringBuilder();
            var solutionContent = new StringBuilder();
            variablesContent.AppendLine($"<{VariablesTag}>");
            solutionContent.AppendLine($"<{SolutionTag}>");

            //
            //Process <see cref="ContextVariables"/> nodes
            //
            context.Log.LogDebug("Processing context variables");
            // Process the context variables nodes
            var stepResults = this.ProcessNodeList(variables, context);
            // Add the context variables to the new plan xml
            variablesContent.Append(stepResults)
                .AppendLine($"</{VariablesTag}>");

            //
            // Process Solution nodes
            //
            context.Log.LogDebug("Processing solution");

            // Process the solution nodes
            stepResults = this.ProcessNodeList(solution, context);
            // Add the solution and context variables updates to the new plan xml
            solutionContent.Append(stepResults)
                .AppendLine($"</{SolutionTag}>");

            // Update the plan xml
            var updatedPlan = goalXmlString + variablesContent + solutionContent;
            updatedPlan = updatedPlan.Trim();
            context.Variables.Update(updatedPlan);

            // Otherwise, execute the next step in the plan
            var nextPlan = (await this._kernel.RunAsync(context.Variables, defaultStepExecutor)).ToString().Trim();

            // And return the updated context with the updated plan xml
            context.Variables.Update(nextPlan);
            return context;
        }
        catch (Exception e) when (!e.IsCriticalException())
        {
            context.Log.LogError(e, "Plan execution failed: {0}", e.Message);
            throw;
        }
    }

    private string ProcessNodeList(XmlNodeList nodeList, SKContext context)
    {
        var stepAndTextResults = new StringBuilder();
        const string INDENT = "  ";
        if (nodeList != null)
        {
            foreach (XmlNode o in nodeList)
            {
                if (o == null)
                {
                    continue;
                }

                var parentNodeName = o.Name;

                context.Log.LogDebug("{0}: found node", parentNodeName);
                foreach (XmlNode o2 in o.ChildNodes)
                {
                    if (o2.Name == "#text")
                    {
                        context.Log.LogDebug("{0}: appending text node", parentNodeName);
                        stepAndTextResults.AppendLine(o2.Value.Trim());
                        continue;
                    }

                    if (o2.Name == StepTag)
                    {
                        context.Log.LogDebug("{0}: appending step node {1}", parentNodeName, o2.OuterXml);
                        stepAndTextResults.Append(INDENT).AppendLine(o2.OuterXml);
                        continue;
                    }

                    stepAndTextResults.Append(INDENT).AppendLine(o2.OuterXml);
                }
            }
        }

        return stepAndTextResults.ToString();
    }

    // TODO: goalTxt is never used
    private static (string goalTxt, string goalXmlString) GatherGoal(XmlDocument xmlDoc)
    {
        XmlNodeList goal = xmlDoc.GetElementsByTagName(GoalTag);
        if (goal.Count == 0)
        {
            throw new PlanningException(PlanningException.ErrorCodes.InvalidPlan, "No goal found.");
        }

        string goalTxt = goal[0]!.FirstChild!.Value ?? string.Empty;
        var goalContent = new StringBuilder();
        goalContent.Append($"<{GoalTag}>")
            .Append(goalTxt)
            .AppendLine($"</{GoalTag}>");
        return (goalTxt, goalContent.ToString());
    }
}

File: python/semantic_kernel/planning/plan_runner.py
# Copyright (c) Microsoft. All rights reserved.
import logging
import xml.etree.ElementTree as ET

from semantic_kernel.kernel_base import KernelBase
from semantic_kernel.orchestration.sk_context import SKContext
from semantic_kernel.orchestration.sk_function import SKFunction
from semantic_kernel.planning.planning_exception import (
    PlanningErrorCode,
    PlanningException,
)


class PlanRunner:
    # The tag name used in the plan xml for the user's goal/ask.
    GOAL_TAG = "goal"
    # The tag name used in the plan xml for the solution.
    SOLUTION_TAG = "solution"
    # The tag name used in the plan xml for a generic step.
    STEP_TAG = "step"
    # The tag name used in the plan xml for the context variables.
    VARIABLES_TAG = "variables"
    # Constants
    INDENT = "  "

    def __init__(self, kernel: KernelBase):
        self._kernel = kernel

    async def execute_xml_plan_async(
        self, context: SKContext, plan_payload: str, default_step_executor: SKFunction
    ):
        try:
            xml_doc = ET.fromstring(f"<xml>{plan_payload}</xml>")

            # Get the Goal
            goal_node, goal_xml_string = self._gather_goal(xml_doc, self.GOAL_TAG)

            # Get the ContextVariables and Solution
            variables = xml_doc.findall(self.VARIABLES_TAG)
            solution = xml_doc.findall(self.SOLUTION_TAG)

            # Prepare content for the new plan xml
            variables_content = []
            solution_content = []
            variables_content.append(f"<{self.VARIABLES_TAG}>")
            solution_content.append(f"<{self.SOLUTION_TAG}>")

            # Process ContextVariables nodes
            logging.debug("Processing context variables")
            # Process the context variables nodes
            step_results = self._process_node_list(variables, context)
            # Add the context variables to the new plan xml
            variables_content.append(step_results)
            variables_content.append(f"</{self.VARIABLES_TAG}>")

            # Process Solution nodes
            logging.debug("Processing solution")
            # Process the solution nodes
            step_results = self._process_node_list(solution, context)
            # Add the solution and context variables updates to the new plan xml
            solution_content.append(step_results)
            solution_content.append(f"</{self.SOLUTION_TAG}>")

            # Update the plan xml
            updated_plan = (
                goal_xml_string + "".join(variables_content) + "".join(solution_content)
            )
            updated_plan = updated_plan.strip()
            context.variables.update(updated_plan)

            # Otherwise, execute the next step in the plan
            next_plan = str(
                await self._kernel.run_async(context.variables, default_step_executor)
            ).strip()

            # And return the updated context with the updated plan xml
            context.variables.update(next_plan)
            return context
        except Exception as e:
            logging.error(f"Plan execution failed: {e}")
            raise

    def _gather_goal(self, xml_doc, GOAL_TAG):
        goal_node = xml_doc.find(GOAL_TAG)
        if goal_node is None:
            raise PlanningException(
                PlanningErrorCode.INVALID_PLAN,
                "Missing goal node in plan xml.",
            )
        return goal_node, ET.tostring(goal_node, encoding="unicode")

    def _process_node_list(self, node_list, context: SKContext):
        step_and_text_results = []
        if node_list is not None:
            for node in node_list:
                if node is None:
                    continue

                parent_node_name = node.tag
                logging.debug(f"{parent_node_name}: found node")
                for child_node in node:
                    if child_node.tag == "#text":
                        logging.debug(f"{parent_node_name}: appending text node")
                        step_and_text_results.append(child_node.text.strip())
                        continue

                    if child_node.tag == self.STEP_TAG:
                        logging.debug(
                            f"{parent_node_name}: appending step node {child_node}"
                        )
                        step_and_text_results.append(
                            f"{self.INDENT}{ET.tostring(child_node, encoding='unicode').strip()}"
                        )
                        continue

                    step_and_text_results.append(
                        f"{self.INDENT}{ET.tostring(child_node, encoding='unicode').strip()}"
                    )
        return "\n".join(step_and_text_results)
